### PR TITLE
Throw exceptions when attempting to use an index whos name is too long

### PR DESCRIPTION
**Pull request by gms8994**
PRing against 5.6 per Taylor.

### CODE REVIEW

File: src/Illuminate/Database/Schema/Grammars/MySqlGrammar.php [hunk 1]
```diff
@@ -215,6 +215,10 @@ public function compileSpatialIndex(Blueprint $blueprint, Fluent $command)
      */
     protected function compileKey(Blueprint $blueprint, Fluent $command, $type)
     {
+        if (strlen($command->index) > 64) {
+            throw new \RunTimeException('Index names longer than 64 characters are not supported by MySQL.');
```
Comment (laurencei) [inline src/Illuminate/Database/Schema/Grammars/MySqlGrammar.php]: Based upon other classes - would be better to use `use RunTimeException` at top of class, then just `throw new RunTimeException (...` here

Same as in the other places this occurs in the other classes?

**Comment by laurencei** [comment]
I dont like the thrown exception that stops the migration.

The problem you will have is up to this point people might have migrations that have been truncating the index on current projects.

If they upgrade to 5.6, then later on try to run a migration - their *old* migrations will fail, and they cant get back to the "starting" point.

i.e. imagine running `migrate:fresh` - it will fail.

So what are their options? They could go and edit an old migration - but that is a big no no generally - you should really only write "new" migrations to move forward. But in this instance they cant.

And if their production database already has a truncated index, but their test environment does not - it might lead to other issues.

I think it should just be a "warning" message on output. People can see this in their development and deicde what action(s) to take. But stopping the migration and not allowing it to move forward is going to cause all sorts of issues for projects that get upgraded.

**Comment by laurencei** [comment]
p.s. I like this PR concept though - because I've been bitten by this before - so being alerted to the problem is definetely a good thing.

**Comment by sisve** [comment]
> The problem you have is up to this point people have migrations that have been truncating the index on current projects.

@laurencei Are you sure this is the current behavior? Mysql, for me, says

    [42000][1059] Identifier name 'long_index_long_index_long_index_long_index_long_index_long_index_long_index_long_index_long_index_l' is too long

The index name is cropped to 100 characters in the error message. But I am already getting exceptions for this case, this PR would just clarify what the limit is.

On the other hand; I do not know what postgresql currently does, or if that I am seeing in mysql is specific to a particular version of mysql.

**Comment by laurencei** [comment]
Postgres definetely silently truncates. I've never seen mySQL throw an error about length, and I know for a fact I've been bitten by truncated index names on mySQL.

Of course - then *if* mySQL throws an error, then you should leave it to the database to handle, and not try to second guess it in PHP.

**Comment by laurencei** [comment]
I know I've been bitten by this before - I just checked an one of my older Laravel projects has a truncated index.

But I agree that *now* mySQL and Postgres seem to throw errors.

In that case - why not just leave it to the database to handle? Why are we second guessing what the database will handle in the migration? Just let it go through to the database and let it decide?

> But I am already getting exceptions for this case, this PR would just clarify what the limit is.
Which brings me back to this PR - if you throw a warning message in the migration, people will see that message before the database error - so you get the best of both worlds.

**Comment by laurencei** [comment]
I know I've been bitten by this before - I just checked one of my older Laravel projects and it has a truncated index for mySQL. But I agree that *now* mySQL seem to throw an error.

But trying it on Postgres on my homestead machine allows the long index through without warning. A quick google implies that Postgres will just happily truncate the index.

So that will be a total breaking change for people in this situation - and its not really "rocoverable" without editing old migrations - which brings me back to the original point.

**Comment by deleugpn** [comment]
Try turning `strict mode` off to see if MySQL will complain about the size of the index. On a counter-note, instead of just a warning, I would suggest to have this configurable. For instance, some people need `Schema::defaultStringLength(191);` in order to run their migration. I know that's not Laravel's fault, but still. What if we introduced `Schema::turnOffIndexValidation()` (awful name, I know, but bear with me) in order to support legacy projects by automatically ignoring the exceptions in case the developer chose to do so?

**Comment by laurencei** [comment]
I tried that. `strict` has no effect.

I found [this SO answer](https://stackoverflow.com/a/28620447/1317935) - which looks like it *was* a mySQL bug that *used* to allow longer than 64 index and foreign keys. It's fixed now - but at least I know I'm not going crazy :)

> I would suggest to have this configurable

Yeah; default to having it "on" - but give people who are hit by this on a legacy project the ability to disable at their own risk. i.e. its an "opt out" feature.

**Comment by mfn** [comment]
We could only throw the exception in development environments.

But for the upgrade scenario pointed out, it's still useless: it prevents the developer from locally working now _unless_ he changes something.

Thing is: if databases decide to *silently* truncate, the probably have the same upgrade path issues we're think about right now.

So if the database is silent, why shouldn't we be? #ShowerThought

If the DB would decide this is unacceptable, the migration would throw an error by itself already.

**Comment by gms8994** [comment]
If the consensus is that this isn't needed, I'm happy to close this and move on to other things.

**Pull request closed without merging**